azkaban-exec-server/src/main/java/azkaban/dag/Dag.java

```diff
@@ -20,27 +20,54 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
  * A DAG (Directed acyclic graph) consists of {@link Node}s.
+ *
+ * <p>Most of the methods in this class should remain package private. Code outside of this
+ * package should mainly interact with the {@link DagService}.
  */
 class Dag {
 
   private final String name;
   private final DagProcessor dagProcessor;
   private final List<Node> nodes = new ArrayList<>();
+  private final Map<String, Node> nameToNodeMap = new HashMap<>();
   private Status status = Status.READY;
 
   Dag(final String name, final DagProcessor dagProcessor) {
+    requireNonNull(name, "The name of the Dag can't be null");
     this.name = name;
     requireNonNull(dagProcessor, "The dagProcessor parameter can't be null.");
     this.dagProcessor = dagProcessor;
   }
 
+  /**
+   * Adds a node to the current dag.
+   *
+   * <p>It's important NOT to expose this method as public. The design relies on this to ensure
```

Comment:
What's the consequence of this node having default scope instead of it being specified as `private`? Is it used elsewhere in the package? Is that intended? Because you mention it NOT being public it might be useful to have a comment saying why it IS default scope.

Reply:
This is how package private access level is specified.

https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

The method is needed by other classes in the same package.